cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -711,6 +719,220 @@ class self_comparator {
 
 }  // namespace equality
 
+namespace hash {
+
+/**
+ * @brief Computes the hash value of an element in the given column.
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <template <typename> class hash_function, typename Nullate>
+class element_hasher {
+ public:
+  __device__ element_hasher(
+    Nullate nulls,
+    uint32_t seed             = DEFAULT_HASH_SEED,
+    hash_value_type null_hash = std::numeric_limits<hash_value_type>::max()) noexcept
+    : _has_nulls(nulls), _seed(seed), _null_hash(null_hash)
+  {
+  }
+
+  template <typename T, CUDF_ENABLE_IF(column_device_view::has_element_accessor<T>())>
+  __device__ hash_value_type operator()(column_device_view const& col,
+                                        size_type row_index) const noexcept
+  {
+    if (_has_nulls && col.is_null(row_index)) { return _null_hash; }
+    return hash_function<T>{_seed}(col.element<T>(row_index));
+  }
+
+  template <typename T, CUDF_ENABLE_IF(not column_device_view::has_element_accessor<T>())>
+  __device__ hash_value_type operator()(column_device_view const& col,
+                                        size_type row_index) const noexcept
+  {
+    cudf_assert(false && "Unsupported type in hash.");
+    return {};
+  }
+
+  uint32_t _seed;
+  hash_value_type _null_hash;
+  Nullate _has_nulls;
+};
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <template <typename> class hash_function, typename Nullate>
+class device_row_hasher {
+  friend class row_hasher;
+
+ public:
+  device_row_hasher() = delete;
+
+  __device__ auto operator()(size_type row_index) const noexcept
+  {
+    // Hash the first column w/ the seed
+    auto const initial_hash =
+      cudf::detail::hash_combine(hash_value_type{0},
+                                 type_dispatcher<dispatch_storage_type>(
+                                   _table.column(0).type(),
```

Comment:
> Can structs contain zero columns?

They sure can, as I recently found out. There's a test for this specific case.

Reply:
My concern was that `_table.column(0)` assumes there are more than zero columns. If this case can be ignored, then don't worry about it. I just wasn't sure if `_table.column(0)` could be an illegal thing to do.

Reply:
I think it can. Based on the assumption that the old code didn't check either. I think if a (internal) user didn't check for the empty column case, they'll get useless results even if we did support this case.